capsules/src/net/udp/driver.rs

```diff
@@ -0,0 +1,473 @@
+                }
+            };
+            let result = self.kernel_tx.map_or(ReturnCode::ENOMEM, |kbuf| {
+                let dst_addr = addr_ports[1].addr;
+                let dst_port = addr_ports[1].port;
+                let src_port = addr_ports[0].port;
+
+                // Copy UDP payload to kernel memory
+                // TODO: handle if too big
+                let result = app.app_write.as_ref().map_or(ReturnCode::ENOMEM, |payload| {
+                        kbuf[..payload.len()].copy_from_slice(payload.as_ref());
+                        self.sender.send_to(dst_addr, dst_port, src_port, &kbuf[..payload.len()])
+                    }
+                );
+                result
+            });
+            if result == ReturnCode::SUCCESS {
+                self.current_app.set(Some(appid));
+            }
+            result
+        })
+    }
+
+    /// Schedule the next transmission if there is one pending. Performs the
+    /// transmission asynchronously, returning any errors via callbacks.
+    #[inline]
+    #[allow(dead_code)]
+    fn do_next_tx_async(&self) {
+        self.get_next_tx_if_idle()
+            .map(|appid| self.perform_tx_async(appid));
+    }
+
+    /// Schedule the next transmission if there is one pending. If the next
+    /// transmission happens to be the one that was just queued, then the
+    /// transmission is synchronous. Hence, errors must be returned immediately.
+    /// On the other hand, if it is some other app, then return any errors via
+    /// callbacks.
+    #[inline]
+    fn do_next_tx_sync(&self, new_appid: AppId) -> ReturnCode {
+        self.get_next_tx_if_idle()
+            .map(|appid| {
+                if appid == new_appid {
+                    self.perform_tx_sync(appid)
+                } else {
+                    self.perform_tx_async(appid);
+                    ReturnCode::SUCCESS
+                }
+            })
+            .unwrap_or(ReturnCode::SUCCESS)
+    }
+
+    #[inline]
+    fn parse_ip_port_pair(&self, buf: &[u8]) -> Option<IPAddrPort> {
+
+        if buf.len() != mem::size_of::<IPAddrPort>() {
+            debug!("[parse] len is {:?}, not {:?} as expected", buf.len(), mem::size_of::<IPAddrPort>());
+            None
+        } else {
+            let (a, p) = buf.split_at(mem::size_of::<IPAddr>());
+            let mut addr = IPAddr::new();
+            addr.0.copy_from_slice(a);
+
+            let pair = IPAddrPort {
+                addr: addr,
+                port: ((p[0] as u16) << 8) + (p[1] as u16),
+            };
+            Some(pair)
+        }
+    }
+}
+
+impl<'a> Driver for UDPDriver<'a> {
+    /// Setup buffers to read/write from.
+    ///
+    /// ### `allow_num`
+    ///
+    /// - `0`: Read buffer. Will contain the received payload.
+    /// - `1`: Write buffer. Contains the UDP payload to be transmitted.
+    /// - `2`: Config buffer. Used to contain miscellaneous data associated with
```

Comment:
This needs to be documented somewhere.

Reply:
now documented in doc/Networking_Userland.md